karate-core/src/main/java/com/intuit/karate/core/StepRuntime.java

```diff
@@ -124,29 +125,45 @@ public String toString() {
     }
 
     private static final Collection<MethodPattern> PATTERNS;
+    private static final Map<String,Collection<Method>> KEYWORDS_METHODS;
 
     static {
         Map<String, MethodPattern> temp = new HashMap();
         List<MethodPattern> overwrite = new ArrayList();
+        KEYWORDS_METHODS = new HashMap<>();
         for (Method method : ScenarioActions.class.getMethods()) {
             When when = method.getDeclaredAnnotation(When.class);
             if (when != null) {
                 String regex = when.value();
-                temp.put(regex, new MethodPattern(method, regex));
+                MethodPattern methodPattern = new MethodPattern(method, regex);
+                temp.put(regex, methodPattern);
+
+                // assuming all @When or @Action start with a ^, get the first word
+                String methodKeyword = regex.substring(1).split(" ")[0];
+                Collection<Method> keywordMethods = KEYWORDS_METHODS.computeIfAbsent(methodKeyword, k -> new HashSet<>());
+                keywordMethods.add(methodPattern.method);
             } else {
                 Action action = method.getDeclaredAnnotation(Action.class);
                 if (action != null) {
                     String regex = action.value();
-                    overwrite.add(new MethodPattern(method, regex));
+                    MethodPattern methodPattern = new MethodPattern(method, regex);
+                    overwrite.add(methodPattern);
+
+                    // assuming all @When or @Action start with a ^, get the first word
+                    String methodKeyword = regex.substring(1).split(" ")[0];
+                    Collection<Method> keywordMethods = KEYWORDS_METHODS.computeIfAbsent(methodKeyword, k -> new HashSet<>());
```

Comment:
now this is a little more code duplication than before. conside moving this logic into the `MethodPattern` object, maybe can be done in the constructor